README.md
# uxt

Unix time helper command

uxt<br />
→946684800

uxt +{int} or -{int}<br />
→946684810

uxt {Unix Epoch}<br />
→Thu May  8 08:13:30 CEST 2025

uxt {Unix Epoch} RFC3339<br />
→2025-05-08T08:13:30+02:00

uxt since {Unix Epoch}<br />
→60 days, 18 hours, 5 minutes, 44 seconds

uxt serv:< port > (local webserver)
```
/current
/plus/<*int>
/minus/<*int>
/since/<*uxt>
/rfc3389/<*uxt>
```

uxt version<br />
→0.0.x

![alt uxt](docs/uxt.gif?raw=true "uxt demo")

CHANGELOG.md
# uxt

## v0.0.4
Add dual function - now with a webservice with the same uxt functionality, ``` serv:<port> ```

## v0.0.3
Add since function, returns time since unixtimestam


## v0.0.2
Change code to split logic from cli parsing

## v0.0.1
Command line tool
* You can add or remove seconds to the current timestamp
* Translate Epoch
* Translate Epoch to RFC3339
* Show current Epoch


serv.go
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package main

import (
	"fmt"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/iobear/uxt/uxt"
)

func handler(w http.ResponseWriter, r *http.Request) {
	path := r.URL.Path
	switch {
	case strings.HasPrefix(path, "/current"):
		fmt.Fprint(w, uxt.GetCurrentUnixTime())
	case strings.HasPrefix(path, "/since/"):
		parts := strings.Split(path, "/")
		if len(parts) != 3 {
			http.Error(w, "Invalid request", http.StatusBadRequest)
			return
		}
		unixTime, err := strconv.ParseInt(parts[2], 10, 64)
		if err != nil {
			http.Error(w, "Invalid timestamp", http.StatusBadRequest)
			return
		}
		fmt.Fprint(w, uxt.GetTimeSince(unixTime))
	case strings.HasPrefix(path, "/convert/"):
		parts := strings.Split(path, "/")
		if len(parts) < 3 {
			http.Error(w, "Invalid request", http.StatusBadRequest)
			return
		}
		unixTime, err := strconv.ParseInt(parts[2], 10, 64)
		if err != nil {
			http.Error(w, "Invalid timestamp", http.StatusBadRequest)
			return
		}
		format := ""
		if len(parts) == 4 {
			format = parts[3]
		}
		result, err := uxt.ConvertUnixTimeToFormattedString(unixTime, format)
		if err != nil {
			http.Error(w, "Error converting timestamp", http.StatusInternalServerError)
			return
		}
		fmt.Fprint(w, result)
	case strings.HasPrefix(path, "/plus/"):
		parts := strings.Split(path, "/")
		if len(parts) != 3 {
			http.Error(w, "Invalid request", http.StatusBadRequest)
			return
		}
		value, err := strconv.Atoi(parts[2])
		if err != nil {
			http.Error(w, "Invalid number", http.StatusBadRequest)
			return
		}
		result, err := uxt.AdjustCurrentUnixTime(value)
		if err != nil {
			http.Error(w, "Error calculating time", http.StatusInternalServerError)
			return
		}
		fmt.Fprint(w, result)
	case strings.HasPrefix(path, "/minus/"):
		parts := strings.Split(path, "/")
		if len(parts) != 3 {
			http.Error(w, "Invalid request", http.StatusBadRequest)
			return
		}
		value, err := strconv.Atoi(parts[2])
		if err != nil {
			http.Error(w, "Invalid number", http.StatusBadRequest)
			return
		}
		result, err := uxt.AdjustCurrentUnixTime(-value) // Note the minus sign here to subtract
		if err != nil {
			http.Error(w, "Error calculating time", http.StatusInternalServerError)
			return
		}
		fmt.Fprint(w, result)
	case strings.HasPrefix(path, "/rfc3339/"):
		parts := strings.Split(path, "/")
		if len(parts) != 3 {
			http.Error(w, "Invalid request", http.StatusBadRequest)
			return
		}
		value, err := strconv.ParseInt(parts[2], 10, 64)
		if err != nil {
			http.Error(w, "Invalid number", http.StatusBadRequest)
			return
		}
		result, err := uxt.ConvertUnixTimeToFormattedString(value, time.RFC3339)
		if err != nil {
			http.Error(w, "Error converting time", http.StatusInternalServerError)
			return
		}
		fmt.Fprint(w, result)
	default:
		http.Error(w, "Not Found", http.StatusNotFound)
	}
}

func RunServer(port string) {
	http.HandleFunc("/", handler)
	http.ListenAndServe(":"+port, nil)
}


uxt/uxt.go
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package uxt

import (
	"fmt"
	"strconv"
	"strings"
	"time"
)

// ConvertUnixTimeToFormattedString converts a Unix timestamp to a formatted string.
func ConvertUnixTimeToFormattedString(unixTime int64, format string) (string, error) {
	t := time.Unix(unixTime, 0)
	var strDate string

	if strings.ToLower(format) == "rfc3339" || format == "3339" {
		strDate = t.Format(time.RFC3339)
	} else {
		strDate = t.Format(time.UnixDate)
	}

	return strDate, nil
}

// AdjustCurrentUnixTime adjusts the current Unix time by the specified number of seconds.
func AdjustCurrentUnixTime(adjustment int) (string, error) {
	now := time.Now().Unix()
	finalTime := now + int64(adjustment)
	return strconv.FormatInt(finalTime, 10), nil
}

// GetCurrentUnixTime returns the current Unix time.
func GetCurrentUnixTime() string {
	return fmt.Sprintf("%v", time.Now().Unix())
}

// GetTimeSince returns the time elapsed since the given Unix timestamp.
func GetTimeSince(unixTime int64) string {
	t := time.Unix(unixTime, 0)
	duration := time.Since(t)

	days := int(duration.Hours()) / 24
	hours := int(duration.Hours()) % 24
	minutes := int(duration.Minutes()) % 60
	seconds := int(duration.Seconds()) % 60

	if days > 0 {
		return fmt.Sprintf("%d days, %d hours, %d minutes, %d seconds", days, hours, minutes, seconds)
	}
	return fmt.Sprintf("%d hours, %d minutes, %d seconds", hours, minutes, seconds)
}


main.go
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"

	"github.com/iobear/uxt/uxt"
)

func main() {
	args := os.Args[1:]
	if len(args) == 0 {
		fmt.Println(uxt.GetCurrentUnixTime())
		return
	}

	switch args[0] {
	case "help", "-h":
		printHelp()
	case "version", "-v":
		printVersion()
	default:
		processUnixTimeArgs(args)
	}
}

func processUnixTimeArgs(args []string) {
	input := args[0]

	if input == "since" {
		if len(args) < 2 {
			printError()
			return
		}
		unixTime, err := strconv.ParseInt(args[1], 10, 64)
		if err != nil {
			printError()
			return
		}

		fmt.Println(uxt.GetTimeSince(unixTime))
		return
	} else if input[0] == '+' || input[0] == '-' {
		adjustment, err := strconv.Atoi(input)
		if err != nil || len(input) < 2 {
			printError()
			return
		}

		result, err := uxt.AdjustCurrentUnixTime(adjustment)
		if err != nil {
			printError()
			return
		}

		fmt.Println(result)
	} else if strings.HasPrefix(input, "serv:") {
		port := strings.TrimPrefix(input, "serv:")
		RunServer(port)
		return
	} else {
		unixTime, err := strconv.ParseInt(input, 10, 64)
		if err != nil {
			printError()
			return
		}

		format := ""
		if len(args) == 2 {
			format = args[1]
		}

		result, err := uxt.ConvertUnixTimeToFormattedString(unixTime, format)
		if err != nil {
			printError()
			return
		}

		fmt.Println(result)
	}
}

func printError() {
	fmt.Println("Invalid argument, try help argument")
}

func printHelp() {
	helpTxt := "This is all the help you get  \n\n"
	helpTxt += os.Args[0] + "\n"
	helpTxt += " 946684800\n\n"
	helpTxt += os.Args[0] + " +<int> or -<int>\n"
	helpTxt += " 946684810\n\n"
	helpTxt += os.Args[0] + " <Unix Epoch> \n"
	helpTxt += " Thu May  8 08:13:30 CEST 2025\n\n"
	helpTxt += os.Args[0] + " <Unix Epoch> RFC3339\n"
	helpTxt += " 2025-05-08T08:13:30+02:00\n\n"
	helpTxt += os.Args[0] + " version\n"
	helpTxt += " 0.0.x\n\n"
	helpTxt += "https://github.com/iobear/uxt \n"

	fmt.Println(helpTxt)
}

func printVersion() {
	fmt.Println("v0.0.4")
}
